### PR TITLE
Scripts/Spells: Druid Travel Form

### PR DESCRIPTION
**Pull request by JayWolf95**
**Changes proposed:**

-  Script for Druid Travel Form

**Target branch(es):** 3.3.5/master

- [x] master

**Issues addressed:** Updates #18840

**Tests performed:** (Does it build, tested in-game, etc.)
- Builds
- Tested in-game.

### CODE REVIEW

File: sql/updates/world/master/9999_99_99_99_world.sql [hunk 1]
```diff
@@ -0,0 +1,7 @@
+DELETE FROM 'spell_script_names' WHERE 'ScriptName' IN ('spell_dru_travel_form');
+INSERT INTO 'spell_script_names' ('spell_id', 'ScriptNames') VALUES
+(783, 'spell_dru_travel_form'),
+(1066, 'spell_dru_travel_form'),
+(783, 'spell_dru_travel_form'),
```
Comment (MartGon) [inline sql/updates/world/master/9999_99_99_99_world.sql]: Duplicated entry on here

**Review by JayWolf95 — COMMENTED** [review]
Fix Duplicate + Add Missing

File: src/server/scripts/Spells/spell_druid.cpp [hunk 2]
```diff
@@ -60,7 +61,128 @@ enum DruidSpells
     SPELL_DRUID_BALANCE_T10_BONUS           = 70718,
     SPELL_DRUID_BALANCE_T10_BONUS_PROC      = 70721,
     SPELL_DRUID_SUNFIRE_DAMAGE              = 164815,
-    SPELL_DRUID_SURVIVAL_INSTINCTS          = 50322
+    SPELL_DRUID_SURVIVAL_INSTINCTS          = 50322,
+    SPELL_DRUID_TRAVEL_FORM                 = 783
+};
```
Comment (MartGon) [inline src/server/scripts/Spells/spell_druid.cpp]: Place this respecting alphabetical order

File: src/server/scripts/Spells/spell_druid.cpp [hunk 3]
```diff
@@ -60,7 +61,128 @@ enum DruidSpells
...
+{
+    DRUID_AQUATIC_FORM                      = 1066,
+    DRUID_FLIGHT_FORM                       = 33943,
+    DRUID_SWIFT_FLIGHT_FORM                 = 40120,
+    DRUID_STAG_FORM                         = 165961
+};
+
+// 783 - Travel Form
+class spell_dru_travel_form : public SpellScriptLoader
+{
```
Comment (MartGon) [inline src/server/scripts/Spells/spell_druid.cpp]: The same for this, try ordering alphabetically by spell_script_name as the header of the file says

File: src/server/scripts/Spells/spell_druid.cpp [hunk 4]
```diff
@@ -1262,6 +1384,7 @@ class spell_dru_wild_growth : public SpellScriptLoader
 
 void AddSC_druid_spell_scripts()
 {
+    new spell_dru_travel_form();
     new spell_dru_dash();
```
Comment (MartGon) [inline src/server/scripts/Spells/spell_druid.cpp]: And in here too

File: src/server/scripts/Spells/spell_druid.cpp [hunk 5]
```diff
@@ -968,6 +978,118 @@ class spell_dru_t4_2p_bonus : public SpellScriptLoader
...
+public:
+    spell_dru_travel_form() : SpellScriptLoader("spell_dru_travel_form") { }
+
+    class spell_dru_travel_form_SpellScript : public SpellScript
+    {
+        PrepareSpellScript(spell_dru_travel_form_SpellScript);
+
+        bool Load()
+        {
+            return GetCaster() && GetCaster()->GetTypeId() == TYPEID_PLAYER;
```
Comment (joschiwald) [inline src/server/scripts/Spells/spell_druid.cpp]: caster never null

File: src/server/scripts/Spells/spell_druid.cpp [hunk 6]
```diff
@@ -968,6 +978,118 @@ class spell_dru_t4_2p_bonus : public SpellScriptLoader
...
+class spell_dru_travel_form : public SpellScriptLoader
+{
+public:
+    spell_dru_travel_form() : SpellScriptLoader("spell_dru_travel_form") { }
+
+    class spell_dru_travel_form_SpellScript : public SpellScript
+    {
+        PrepareSpellScript(spell_dru_travel_form_SpellScript);
+
+        bool Load()
```
Comment (Keader) [inline src/server/scripts/Spells/spell_druid.cpp]: missing override

File: src/server/scripts/Spells/spell_druid.cpp [hunk 7]
```diff
@@ -968,6 +978,118 @@ class spell_dru_t4_2p_bonus : public SpellScriptLoader
...
+        {
+            return GetCaster()->GetTypeId() == TYPEID_PLAYER;
+        }
+
+        SpellCastResult CheckCast()
```
Comment (Keader) [inline src/server/scripts/Spells/spell_druid.cpp]: missing override

Comment (JayWolf95) [inline src/server/scripts/Spells/spell_druid.cpp]: Does CheckCast() need override? None of the other CheckCast() declarations in this script use override.

Comment (joschiwald) [inline src/server/scripts/Spells/spell_druid.cpp]: no, it does not override anything

File: src/server/scripts/Spells/spell_druid.cpp [hunk 8]
```diff
@@ -968,6 +978,118 @@ class spell_dru_t4_2p_bonus : public SpellScriptLoader
...
+        PrepareAuraScript(spell_dru_travel_form_AuraScript);
+
+        void AfterApply(AuraEffect const* /*aurEff*/, AuraEffectHandleModes /*mode*/)
+        {
+            Player* player = GetTarget()->ToPlayer();
+            if (player->IsInWater())
+                player->CastSpell(player, DRUID_AQUATIC_FORM, true);
+            else if (player->IsInCombat())
+                player->CastSpell(player, DRUID_STAG_FORM, true);
+            else if ((player->GetMapId() == 530) || (player->GetMapId() == 571 && player->HasSpell(54197)) || ((player->GetMapId() == 0 || player->GetMapId() == 1) && player->HasSpell(90267)))
```
Comment (Aokromes) [inline src/server/scripts/Spells/spell_druid.cpp]: this can't be done on a generic way?

Comment (JayWolf95) [inline src/server/scripts/Spells/spell_druid.cpp]: I'm not sure, I've looked at player->CanFlyInZone(), and it seems outdated, but I could be wrong.

**Comment by DoctorKraft** [comment]
It's totally wrong and not blizzlike, the travel form are updated with environnement for example you flying (raven form) and after you go in water, travel form are automatically updated in water form.

**Comment by JayWolf95** [comment]
@DoctorKraft Do you have any suggestions on how to implement that? I have a previous WIP somewhere, I'd have to find it, but my attempt caused a infinite loop.

**Comment by DoctorKraft** [comment]
@JayWolf95 I have made this https://github.com/TrinityCore/TrinityCore/pull/19073 with playerscript and working well but yeah the method are shitty because each movement the script are called for check if form are good with environnement. Maybe create a function cached if no change in environnement don't call aways the script

**Comment by ArcaneFox** [comment]
I think the best way to handle this:


1) Create a new event in the Player: OnEnviromentChanged
(https://msdn.microsoft.com/en-us/library/aa645739(v=vs.71).aspx)

2) Create a new method in the Player
```C#
public void ChangeEnviroment(EnviromentType et) {
	switch(et)
		case Water:
			SetInWater(true);
			break;
		
		case Ground:
			SetInWater(false);
			break;
			
		// case lava, etc
			
		default:
			// log error - unknown enviroment type
			break;
	
	// call every subscribed
	OnEnviromentChanged();
}
```

3) Subscribe to the OnEnviromentChanged in the spell_druid.cpp spell_dru_travel_form

4) Change access level to private of the SetInWater method

File: src/server/scripts/Spells/spell_druid.cpp [hunk 9]
```diff
@@ -968,6 +978,118 @@ class spell_dru_t4_2p_bonus : public SpellScriptLoader
...
+                {
+                    player->RemoveAura(SPELL_DRUID_TRAVEL_FORM, ObjectGuid::Empty, 0U, AURA_REMOVE_BY_CANCEL);
+                    player->RemoveAura(DRUID_STAG_FORM, ObjectGuid::Empty, 0U, AURA_REMOVE_BY_CANCEL);
+                    player->RemoveAura(DRUID_AQUATIC_FORM, ObjectGuid::Empty, 0U, AURA_REMOVE_BY_CANCEL);
+                    player->RemoveAura(DRUID_FLIGHT_FORM, ObjectGuid::Empty, 0U, AURA_REMOVE_BY_CANCEL);
+                    player->RemoveAura(DRUID_SWIFT_FLIGHT_FORM, ObjectGuid::Empty, 0U, AURA_REMOVE_BY_CANCEL);
+                }
+                else if ((player->GetMapId() == 530) || (player->GetMapId() == 571 && player->HasSpell(54197)) || ((player->GetMapId() == 0 || player->GetMapId() == 1) && player->HasSpell(90267)))
+                {
+                    if (player->getLevel() >= 71)
```
Comment (Mihapro) [inline src/server/scripts/Spells/spell_druid.cpp]: Is there no other requirements for Swift Flight Form anymore?
http://wowwiki.wikia.com/wiki/Swift_Flight_Form_quest_chain

**Comment by Shauren** [comment]
replaced by #21046

**Pull request closed without merging**